edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/brokerConnection/MqttBrokerConnector.cs

```diff
@@ -0,0 +1,423 @@
+    using uPLibrary.Networking.M2Mqtt.Messages;
+
+    public class MqttBrokerConnector : IMqttBrokerConnector
+    {
+        const int ReconnectDelayMs = 2000;
+
+        readonly IComponentDiscovery components;
+        readonly ISystemComponentIdProvider systemComponentIdProvider;
+        readonly Dictionary<ushort, TaskCompletionSource<bool>> pendingAcks = new Dictionary<ushort, TaskCompletionSource<bool>>();
+
+        readonly object guard = new object();
+
+        Option<Channel<MqttPublishInfo>> publications;
+        Option<Task> forwardingLoop;
+        Option<MqttClient> mqttClient;
+
+        public MqttBrokerConnector(IComponentDiscovery components, ISystemComponentIdProvider systemComponentIdProvider)
+        {
+            this.components = Preconditions.CheckNotNull(components);
+            this.systemComponentIdProvider = Preconditions.CheckNotNull(systemComponentIdProvider);
+
+            // because of the circular dependency between MqttBridgeConnector and the producers,
+            // in this loop the producers get the IMqttBridgeConnector reference:
+            foreach (var producer in components.Producers)
+            {
+                producer.SetConnector(this);
+            }
+        }
+
+        public async Task ConnectAsync(string serverAddress, int port)
+        {
+            Events.Starting();
+
+            var client = default(MqttClient);
+
+            lock (this.guard)
+            {
+                if (this.mqttClient.HasValue)
+                {
+                    Events.StartedWhenAlreadyRunning();
+                    throw new InvalidOperationException("Cannot start mqtt-bridge connector twice");
+                }
+
+                client = new MqttClient(serverAddress, port, false, MqttSslProtocols.None, null, null);
+                this.mqttClient = Option.Some(client);
+            }
+
+            client.MqttMsgPublished += this.ConfirmPublished;
+            client.MqttMsgPublishReceived += this.ForwardPublish;
+
+            this.publications = Option.Some(Channel.CreateUnbounded<MqttPublishInfo>(
+                                    new UnboundedChannelOptions
+                                    {
+                                        SingleReader = true,
+                                        SingleWriter = true
+                                    }));
+
+            this.forwardingLoop = Option.Some(this.StartForwardingLoop());
+
+            // if ConnectAsync is supposed to manage starting it with broker down,
+            // put a loop here to keep trying - see 'TriggerReconnect' below
+            var isConnected = await TryConnectAsync(client, this.components.Consumers, this.systemComponentIdProvider.EdgeHubBridgeId);
+
+            if (!isConnected)
+            {
+                client.MqttMsgPublished -= this.ConfirmPublished;
+                client.MqttMsgPublishReceived -= this.ForwardPublish;
+
+                await this.StopForwardingLoopAsync();
```

Comment:
I feel we can simplify this logic quite a bit by making this class non-resumable. The protocol head can just create a new connector if it wants to restart it.